dbt_platform_helper/domain/codebase.py

```diff
@@ -0,0 +1,298 @@
+                    {"name": "IMAGE_TAG", "value": f"commit-{commit}"},
+                ],
+            },
+        )
+
+        if build_url:
+            return self.echo_fn(
+                "Your deployment has been triggered. Check your build progress in the AWS Console: "
+                f"{build_url}",
+            )
+
+        return self.echo_fn("Your deployment was not triggered.")
+
+    def list(self, app: str, with_images: bool):
+        """List available codebases for the application."""
+        session = self.get_aws_session_or_abort_fn()
+        application = self.__load_application_or_abort(session, app)
+        ssm_client = session.client("ssm")
+        ecr_client = session.client("ecr")
+        parameters = ssm_client.get_parameters_by_path(
+            Path=f"/copilot/applications/{application.name}/codebases",
+            Recursive=True,
+        )["Parameters"]
+
+        codebases = [json.loads(p["Value"]) for p in parameters]
+
+        if not codebases:
+            self.echo_fn(f'No codebases found for application "{application.name}"', fg="red")
+            raise SystemExit(1)
+
+        self.echo_fn("The following codebases are available:")
+
+        for codebase in codebases:
+            self.echo_fn(f"- {codebase['name']} (https://github.com/{codebase['repository']})")
+            if with_images:
+                self._list_latest_images(
+                    ecr_client, f"{application.name}/{codebase['name']}", codebase["repository"]
+                )
+
+        self.echo_fn("")
+
+    def __get_build_url_from_arn(self, build_arn: str) -> str:
+        _, _, _, region, account_id, project_name, build_id = build_arn.split(":")
+        project_name = project_name.removeprefix("build/")
+        return (
+            f"https://eu-west-2.console.aws.amazon.com/codesuite/codebuild/{account_id}/projects/"
+            f"{project_name}/build/{project_name}%3A{build_id}"
+        )
+
+    def __load_application_or_abort(self, session: Session, app: str) -> Application:
+        try:
+            return self.load_application_fn(app, default_session=session)
+        except ApplicationNotFoundError:
+            self.echo_fn(
+                f"""The account "{os.environ.get("AWS_PROFILE")}" does not contain the application "{app}"; ensure you have set the environment variable "AWS_PROFILE" correctly.""",
+                fg="red",
+            )
+            raise click.Abort
+
+    def __load_application_with_environment(self, session: Session, app, env):
+        application = self.__load_application_or_abort(session, app)
+
+        if not application.environments.get(env):
+            self.echo_fn(
+                f"""The environment "{env}" either does not exist or has not been deployed.""",
+                fg="red",
+            )
+            raise click.Abort
+        return application
+
+    def __check_codebase_exists(self, session: Session, application: Application, codebase: str):
```

Comment:
This seems like AWS based funcitonality does this belong in a provider?

Reply:
Yeah I think ultimately this should be in an AWS provider - or similar - that gets injected. For now we've followed what was in the Database Copy domain class. I'd be minded to say this could be updated in a subsequent refactor, though, especially if you and Emile are working on an AWS provider as part of your work? Could move this code into there once you're done with your ticket

Reply:
I think that is reasonable, I wouldn't worry about trying to work out what provider classes we have right now. Although do you think there would still be value in refactoring this out into into the provider like `datacopy` did with the load balancer helper functions?
Currently for the conduit work we have done that too, where the initial refactor is only moving the helper functions into its own provider file.

Reply:
Ah right, I get you. Yeah, I don't think it should be too much to move these functions out in a similar way. Then it'll be a smaller increment to refactor into a class if that's the intention later down the line